Lib/asyncio/proactor_events.py

```diff
@@ -426,6 +427,120 @@ def _pipe_closed(self, fut):
             self.close()
 
 
+class _ProactorDatagramTransport(_ProactorBasePipeTransport):
+
+    def __init__(self, loop, sock, protocol, address=None,
+                 waiter=None, extra=None):
+        super().__init__(loop, sock, protocol, waiter=waiter, extra=extra)
+        self._address = address
+        # We don't need to call _protocol.connection_made() since our base
+        # constructor does it for us.
+        self._buffer = collections.deque()
+        self._loop.call_soon(self._loop_reading)
+
+    def _set_extra(self, sock):
+        super()._set_extra(sock)
+        self._extra['socket'] = sock
+
+        try:
+            self._extra['sockname'] = sock.getsockname()
+        except (socket.error, AttributeError):
+            if self._loop.get_debug():
+                logger.warning(
+                    "getsockname() failed on %r", sock, exc_info=True)
+
+        if 'peername' not in self._extra:
+            try:
+                self._extra['peername'] = sock.getpeername()
+            except (socket.error, AttributeError):
+                if self._loop.get_debug():
+                    logger.warning("getpeername() failed on %r",
+                                   sock, exc_info=True)
+
+
+    def abort(self):
+        self._force_close(None)
+
+    def sendto(self, data, addr=None):
+        if not isinstance(data, (bytes, bytearray, memoryview)):
+            raise TypeError('data argument must be byte-ish (%r)',
+                            type(data))
+
+        if not data:
+            return
+
+        if self._conn_lost and self._address:
+            # close() or force_close() has been called on the bound endpoint
```

Comment:
Again, _SelectorDatagramTransport has a different code:

            if self._conn_lost >= constants.LOG_THRESHOLD_FOR_CONNLOST_WRITES:
                logger.warning('socket.send() raised exception.')
            self._conn_lost += 1

Please use that.

Note: _SelectorDatagramTransport.sendto() error message should use "sendto()" rather than "socket.send()".